### PR TITLE
internal/transport: fix severity of log when receiving a GOAWAY with error code ENHANCE_YOUR_CALM

### LINKED ISSUE #5903: transport: fix severity of log when receiving a GOAWAY with error code ENHANCE_YOUR_CALM
According to [A8](https://github.com/grpc/proposal/blob/master/A8-client-side-keepalive.md#basic-keepalive)

```
When a client receives a GOAWAY with error code ENHANCE_YOUR_CALM and debug 
data equal to ASCII "too_many_pings", it should log the occurrence at a log level that is 
enabled by default and double the configure KEEPALIVE_TIME used for new connections
on that channel.
```

But we are logging this at `INFO` and verbosity level `2`. https://github.com/grpc/grpc-go/blob/12b8fb52a18c8a1667dde7a4f8087ecdd2abbeaf/internal/transport/http2_client.go#L1258

### PR DESCRIPTION
**Pull request by hnlq715**
This pr fix #5903

RELEASE NOTES: N/A

### CODE REVIEW

File: internal/transport/http2_client.go [hunk 1]
```diff
@@ -1265,7 +1265,11 @@ func (t *http2Client) handleGoAway(f *http2.GoAwayFrame) {
 		return
 	}
 	if f.ErrCode == http2.ErrCodeEnhanceYourCalm {
```
Comment (zasweq) [inline internal/transport/http2_client.go:1267]: Perhaps add the too many pings conditional here, corresponding to the and in this line: with error code ENHANCE_YOUR_CALM "and" debug data equal to ASCII "too_many_pings"

File: internal/transport/http2_client.go [hunk 2]
```diff
@@ -1265,7 +1265,11 @@ func (t *http2Client) handleGoAway(f *http2.GoAwayFrame) {
...
-		if logger.V(logLevel) {
+		// When a client receives a GOAWAY with error code ENHANCE_YOUR_CALM and debug
+		// data equal to ASCII "too_many_pings", it should log the occurrence at a log level that is
+		// enabled by default and double the configure KEEPALIVE_TIME used for new connections
+		// on that channel.
+		if string(f.DebugData()) == "too_many_pings" {
 			logger.Infof("Client received GoAway with http2.ErrCodeEnhanceYourCalm.")
```
Comment (zasweq) [inline internal/transport/http2_client.go:1273]: Switch to default log level, not Info, and also perhaps add the doubling of keep alive time here?

**Review by zasweq — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: internal/transport/http2_client.go (hunk 2 above)
Comment (zasweq) [reply internal/transport/http2_client.go:1273]: Btw, if you switch this to error, some tests might fail, as the test runner fails on seeing any unexpected error logs.

File: internal/transport/http2_client.go (hunk 1 above)
Comment (hnlq715) [reply internal/transport/http2_client.go:1267]: done

File: internal/transport/http2_client.go (hunk 2 above)
Comment (hnlq715) [reply internal/transport/http2_client.go]: Got it, thanks

**Comment by zasweq** [comment]
Btw, I only see testing changes for this PR since my last review. I had a comment about the implementation itself. Once you get to that, I can take a look.

**Comment by hnlq715** [comment]
@zasweq Sorry for the delay.

With Adding `grpctest.TLogger.ExpectError` to fix test, here is the data race warning when running go test with race.
I would try to figure out.

```
WARNING: DATA RACE
Read at 0x00c000708383 by goroutine 3453:
  testing.(*common).logDepth()
      /opt/hostedtoolcache/go/1.19.5/x64/src/testing/testing.go:883 +0xc4
  testing.(*common).log()
      /opt/hostedtoolcache/go/1.19.5/x64/src/testing/testing.go:876 +0x84
  testing.(*common).Log()
      /opt/hostedtoolcache/go/1.19.5/x64/src/testing/testing.go:917 +0x5b
  google.golang.org/grpc/internal/grpctest.(*tLogger).log()
      /home/runner/work/grpc-go/grpc-go/internal/grpctest/tlogger.go:116 +0x4a9
  google.golang.org/grpc/internal/grpctest.(*tLogger).WarningDepth()
      /home/runner/work/grpc-go/grpc-go/internal/grpctest/tlogger.go:[226](https://github.com/grpc/grpc-go/actions/runs/4035389242/jobs/6937256777#step:8:227) +0x64
  google.golang.org/grpc/internal/grpclog.WarningDepth()
      /home/runner/work/grpc-go/grpc-go/internal/grpclog/grpclog.go:44 +0x85
  google.golang.org/grpc/grpclog.(*componentData).WarningDepth()
      /home/runner/work/grpc-go/grpc-go/grpclog/component.go:41 +0x16b
  google.golang.org/grpc/internal/grpclog.(*PrefixLogger).Warningf()
      /home/runner/work/grpc-go/grpc-go/internal/grpclog/prefixLogger.go:48 +0x16c
  google.golang.org/grpc/xds/internal/xdsclient.(*authority).handleWatchTimerExpiry()
      /home/runner/work/grpc-go/grpc-go/xds/internal/xdsclient/authority.go:398 +0x1b6
  google.golang.org/grpc/xds/internal/xdsclient.(*authority).watchResource.func1()
      /home/runner/work/grpc-go/grpc-go/xds/internal/xdsclient/authority.go:359 +0x7a
```

**Comment by hnlq715** [comment]
And, seems like this is a common race case like https://github.com/grpc/grpc-go/issues/5956

**Comment by zasweq** [comment]
Yes, please fix the test and get to my implementation comment :).

**Comment by hnlq715** [comment]
> Yes, please fix the test and get to my implementation comment :).

I've changed the log level and add too many pings in log message.

But the failed test seems related to the test log framework, and it would take times to figure out.

**Comment by hnlq715** [comment]
BTW, this problem is also difficult to reproduce in my local environment.

**Review by zasweq — CHANGES_REQUESTED** [review]
Hey thanks, the flake you linked was actually an orthogonal flaky issue. Some minor nits but looks good ugly.

File: internal/transport/http2_client.go [hunk 3]
```diff
@@ -1265,8 +1265,12 @@ func (t *http2Client) handleGoAway(f *http2.GoAwayFrame) {
 		return
 	}
 	if f.ErrCode == http2.ErrCodeEnhanceYourCalm {
-		if logger.V(logLevel) {
-			logger.Infof("Client received GoAway with http2.ErrCodeEnhanceYourCalm.")
+		// When a client receives a GOAWAY with error code ENHANCE_YOUR_CALM and debug
+		// data equal to ASCII "too_many_pings", it should log the occurrence at a log level that is
+		// enabled by default and double the configure KEEPALIVE_TIME used for new connections
+		// on that channel.
+		if string(f.DebugData()) == "too_many_pings" {
```
[the remark this answers is not in the document]
Comment (hnlq715) [reply internal/transport/http2_client.go]: done :-)

File: internal/transport/http2_client.go [hunk 4]
```diff
@@ -1265,8 +1265,12 @@ func (t *http2Client) handleGoAway(f *http2.GoAwayFrame) {
...
+			logger.Errorf("Client received GoAway with http2.ErrCodeEnhanceYourCalm and debug data equal to ASCII \"too_many_pings\".")
```
[the remark this answers is not in the document]
Comment (hnlq715) [reply internal/transport/http2_client.go]: done :-)

**Review by zasweq — APPROVED** [review]
LGTM. Thanks for the PR. This needs two approvals, so handing off to my TL Doug.

**Review by dfawley — APPROVED** [review]

**Comment by dfawley** [comment]
@zasweq please re-add your approval and merge.  Thanks!

**Pull request merged**